extensions/amp-story/1.0/amp-story-page.js

```diff
@@ -322,20 +327,163 @@ export class AmpStoryPage extends AMP.BaseElement {
+
+    if (this.getAmpDoc().isVisible()) {
+      this.setupAutoAdvanceForVisible_();
+      this.maybeSetStoryNextUp_();
+      if (this.isActive() && handlePreviewToVisibleTransition) {
+        this.handlePreviewToVisibleTransition_();
+      } else {
+        this.initializeAdvancementConfig_();
+      }
+    }
+  }
+
+  /**
+   * Configures the page to auto advance using preview-specific durations.
+   * @private
+   */
+  setupAutoAdvanceForPreview_() {
+    let previewSecondsPerPage = this.viewer_.getParam('previewSecondsPerPage');
+    if (previewSecondsPerPage) {
+      previewSecondsPerPage = parseInt(previewSecondsPerPage, 10);
+    }
+    if (isNaN(previewSecondsPerPage) || previewSecondsPerPage <= 0) {
+      previewSecondsPerPage = DEFAULT_PREVIEW_AUTO_ADVANCE_DURATION_S;
+    }
+    this.element.setAttribute(
+      'auto-advance-after',
+      previewSecondsPerPage + 's'
+    );
+
+    const firstVideo = this.getFirstAmpVideo_();
+    if (firstVideo) {
+      whenUpgradedToCustomElement(firstVideo)
+        .then(() => firstVideo.getImpl())
+        .then((videoImpl) => {
+          const loopVideoIfTooShort = (vidLength) => {
+            const videoEl = firstVideo.querySelector('video');
+            const tooShort = vidLength < previewSecondsPerPage;
+            videoEl.loop ||= tooShort;
+          };
+
+          if (videoImpl.getDuration()) {
+            loopVideoIfTooShort(videoImpl.getDuration());
+          } else {
+            listenOnce(firstVideo, VideoEvents_Enum.LOADEDMETADATA, () => {
+              loopVideoIfTooShort(videoImpl.getDuration());
+            });
+          }
+        });
+    }
+  }
+
+  /**
+   * Configures the page to auto advance using default durations.
+   * @private
+   */
+  setupAutoAdvanceForVisible_() {
+    // The 'auto-advance-after' attribute value may have been altered if auto
+    // advance was set up for preview mode. We revert this alteration for
+    // visible mode, in accordance with any values specified by the publisher.
+    if (this.initialAutoAdvanceValue_) {
+      this.element.setAttribute(
+        'auto-advance-after',
+        this.initialAutoAdvanceValue_
+      );
+    } else {
+      this.element.removeAttribute('auto-advance-after');
+    }
+  }
+
+  /**
+   * Handles the transition between the `preview` and `visible` visibility
+   * states by reinitializing this page's advancement config.
+   *
+   * As long as this page has called `setupAutoAdvanceForVisible_()`, prior to
+   * this method, then the reinitialization will result in a new advancement
+   * config that aligns with the advancement logic specified by the publisher.
+   * @private
+   */
+  handlePreviewToVisibleTransition_() {
+    devAssert(
+      this.advancement_?.getType() === 'TimeBasedAdvancement',
+      'The advancement is expected to be time-based in preview mode'
+    );
+
+    const progress = this.advancement_.getProgress();
+    const progressMs = this.advancement_.getProgressMs();
```

Comment:
You're fetching these two values but only using one at any time, would make more sense to fetch them in-line

Reply:
The two values are fetched here because the `this.initializeAdvancementConfig_()` call beneath will result in a new advancement, so we can't fetch them inline. I've added a comment to clarify this.